client/write.go

```diff
@@ -151,24 +154,26 @@ func (w *writeState) completionFn(result interface{}, err error) {
 		w.errors = append(w.errors, fmt.Errorf("missing shard %d in host %s", w.op.shardID, hostID))
 	} else if shardState == shard.Available {
 		// NB(bl): only count writes to available shards towards success
-		w.success++
+		addToSuccess = 1
 	}
+	success := atomic.AddInt32(&w.success, addToSuccess)
 
+	w.Lock()
 	switch w.session.writeLevel {
 	case topology.ConsistencyLevelOne:
-		if w.success > 0 || w.pending == 0 {
+		if success > 0 || pending == 0 {
 			w.Signal()
 		}
 	case topology.ConsistencyLevelMajority:
-		if w.success > w.majority || w.pending == 0 {
+		if success >= w.majority || pending == 0 {
```

Comment:
Can we add a test to make sure we catch this in the future? (presumably it fails with the master but passes with this change)